Implement the "dotNetCliPaths" option to support custom .NET SDK locations

jkoritzinsky opened the pull request:
Depends on https://github.com/OmniSharp/omnisharp-roslyn/pull/2227

SamB commented:
... don't you need to register the option in package.json ?

jkoritzinsky commented:
@SamB I did not realize I needed to do that. Done!

JoeRobich commented:
@jkoritzinsky The c# extension also performs a check for `dotnet` to be on the PATH. Now that we are adding the ability to specify additional paths, can you update the [getDotnetInfo](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/utils/getDotnetInfo.ts#L16) function to return results when `dotnet` isn't on the path but instead specified by the new setting.

gregg-miskelly commented:
The debugger also needs to find `dotnet`. So this directory probably needs to be added to PATH through `DebugAdapterExecutable.options.env` in [`createDebugAdapterDescriptor`](https://github.com/OmniSharp/omnisharp-vscode/blob/1756e252a74b051097ed9acb9e080fba8fdff185/src/coreclr-debug/activate.ts#L180).

jkoritzinsky commented:
@gregg-miskelly @JoeRobich I've finally had a chance to get back to this PR. I've updated it to master and implemented both of your requests. Can you take another look when you have a chance?

JoeRobich commented on src/features/commands.ts:
```diff
@@ -53,7 +53,7 @@ export default function registerCommands(context: vscode.ExtensionContext, serve
     disposable.add(vscode.commands.registerCommand('dotnet.generateAssets', async (selectedIndex) => generateAssets(server, selectedIndex)));
 
-    disposable.add(vscode.commands.registerCommand('csharp.reportIssue', async () => reportIssue(vscode, eventStream, getDotnetInfo, platformInfo.isValidPlatformForMono(), optionProvider.GetLatestOptions(), monoResolver)));
+    disposable.add(vscode.commands.registerCommand('csharp.reportIssue', async () => reportIssue(vscode, eventStream, getDotnetInfo, platformInfo.isValidPlatformForMono(), optionProvider.GetLatestOptions(), monoResolver, dotnetResolver, )));
```
nit: trailing comma in parameters

```suggestion
    disposable.add(vscode.commands.registerCommand('csharp.reportIssue', async () => reportIssue(vscode, eventStream, getDotnetInfo, platformInfo.isValidPlatformForMono(), optionProvider.GetLatestOptions(), monoResolver, dotnetResolver)));
```

JoeRobich commented on src/utils/getDotnetInfo.ts:
```diff
@@ -13,15 +15,28 @@ let _dotnetInfo: DotnetInfo;
-        let data = await execChildProcess('dotnet --info', process.cwd());
+        let data = await execChildProcess(`${dotnetExecutablePath || 'dotnet'} --info`, process.cwd());
+
+        dotnetInfo.CliPath = dotnetExecutablePath; 
```
Just wanted to ensure that the intention is this will be `undefined` when falling back to running `dotnet` from PATH.

jkoritzinsky replied:
Yes, this is intentional.

JoeRobich reviewed (commented):
@jkoritzinsky Thanks for sticking with this PR!

JoeRobich commented on src/coreclr-debug/activate.ts:
```diff
@@ -172,12 +173,13 @@ export class DebugAdapterExecutableFactory implements vscode.DebugAdapterDescrip
         if (!executable) {
             const command = path.join(common.getExtensionPath(), ".debugger", targetArchitecture, "vsdbg-ui" + CoreClrDebugUtil.getPlatformExeExtension());
-            executable = new vscode.DebugAdapterExecutable(command);
+            executable = new vscode.DebugAdapterExecutable(command, [], { env: { 'DOTNET_ROOT' : dotNetInfo.CliPath } });
```
Since `CliPath` is the full path to the alternate dotnet CLI, I believe we should be parsing out the containing folder path when setting DOTNET_ROOT.

jkoritzinsky commented:
@JoeRobich @filipw can one of you merge this PR for me? Or is there something else I need to do before we get this merged in?

JoeRobich commented:
@jkoritzinsky Sorry for the delay. Thanks again!
